### PR TITLE
Feature/ci upload artifacts to owncloud 3

### PR DESCRIPTION
**Pull request by rmelotte**
As mentioned during the sync meeting, I changed so many things from https://github.com/prplfoundation/prplMesh/pull/1050 that I decided to create a whole new branch and PR.

I think I addressed all review comments, and made it so that artifacts are uploaded to owncloud for every branch (first in a temporary folder, then moving them into their final location).
When on the master branch, an additional copy is made to the "latest" folder.

The artifacts URL has also changed, as the directory is now back to `prplmesh/artifacts` (I figured out how to add it to the robot account home).

### CODE REVIEW

File: ci/upload_to_owncloud.sh [hunk 1]
```diff
@@ -53,11 +53,16 @@ main() {
...
-            realpath {} --relative-to="$(dirname "$local_path")" \; | \
+            realpath {} --relative-to="$(dirname "$local_path")" \; | {
+                error=0
                 while read -r -s dir; do
                     printf . # show progress
                     dbg "Create directory: $dir"
-                    curl ${QUIET:+ -s -S} -f -n -X MKCOL "$OWNCLOUD_URL/$user/$remote_path/$dir"
+                    curl ${QUIET:+ -s -S} -f -n -X MKCOL "$OWNCLOUD_URL/$user/$remote_path/$dir" || {
+                        error="$?"
+                        err "Failed to create dir: $remote_path/$dir/ (error $error) skipping."
```
Comment (tomereli) [inline ci/upload_to_owncloud.sh]: It depends on the error - but if the directory exists we might still want to upload the files, don't we? In this case, perhaps we should not `continue`.

File: tools/docker/builder/Dockerfile [hunk 2]
```diff
@@ -32,6 +32,7 @@ RUN apt-get update && apt-get install -y \
     python-yaml \
     python3 \
     python3-yaml \
+    uuid-runtime \
```
Comment (tomereli) [inline tools/docker/builder/Dockerfile:35]: I assume this should not only be used by docker - so please add it to the main readme as well

File: ci/upload_to_owncloud.sh [hunk 3]
```diff
@@ -137,12 +137,29 @@ main() {
     local_path="$1"; shift
 
     info "upload $local_path to $OWNCLOUD_BROWSE_URL/$remote_path/$(basename "$local_path") (using user $user)"
+
+    if ! command -v uuidgen > /dev/null ; then
```
Comment (tomereli) [inline ci/upload_to_owncloud.sh:61]: So now we always upload to a temporary location before moving it to the final location? Isn't it wastefull? Uploading a full logs diretocy of 20+ tests takes A LOT of time as it is, I think this feature should be optional unless you tested it with a big folder with lots of files and found out that the move part is significantly faster than the upload part.

**Review by tomereli — APPROVED** [review]
One major comment about the potentially very long time to upload log folders - so please test that the move operation is negligible or make this optional.

File: ci/upload_to_owncloud.sh (hunk 1 above)
Comment (rmelotte) [reply ci/upload_to_owncloud.sh]: If the directory already exists it means something bad happened, since we're always uploading to a temporary folder that is supposed to be empty.

Comment (tomereli) [reply ci/upload_to_owncloud.sh]: I would like to revisit this - when used by the `run_easymesh_cert` it is practically guaranteed that the directory will not exist so the whole temporary folder solution is not needed in this case which is why I suggest to make it optional.

File: ci/upload_to_owncloud.sh (hunk 3 above)
Comment (rmelotte) [reply ci/upload_to_owncloud.sh:61]: Yes we always upload to a temporary location first, then do a move.

If the destination folder doesn't exist the move is instant since it's just a rename, so there shouldn't be any significant overhead.

When the folder already exists however (and it does for the `latest` folder), it takes much more time.
If we didn't upload to a temporary folder first, we would anyway have to remove the destination folder first before uploading, so I'm not convinced it would take less time.


To summarize: 
I think the move is annoying for the `latest` folder, because I assumed it would almost instantly replace the folder (we want it to contain valid artifacts at any time) and it doesn't, but it has nothing to do with the fact that we use temporary folders.
To improve this we could try [Frederik's suggestion](https://github.com/prplfoundation/prplMesh/pull/1050#discussion_r400861726) and see if we can use some kind of link instead of a plain folder.

Comment (arnout) [reply ci/upload_to_owncloud.sh:61]: I like Frederik's suggestion. So with that, we could say that we *require* the target folder not to exist, right?

That said, I think we should merge it as is now, even if the copy takes some time, and fix later. For Tomer's use case the temporary location shouldn't hurt, and this way we at least have something working for the "latest".

File: ci/upload_to_owncloud.sh [hunk 4]
```diff
@@ -29,6 +29,52 @@ usage() {
...
+    #   OWNCLOUD_URL
+    #   user
+    local dir delete status
+    dir="$1"
+    delete="$2"
+    status=0
+
+    dbg "Create directory: $dir"
+    if [ "$delete" = true ] ; then
+        curl ${QUIET:+-s -S} -f -n -X DELETE "$OWNCLOUD_URL/$user/$dir" > /dev/null 2>&1
```
Comment (arnout) [inline ci/upload_to_owncloud.sh]: Given the number of times `curl ${QUIET:+-s -S} -f -n -X ` is repeated, it might make sense to make a wrapper for that as well (nitpick).

File: tools/docker/builder/Dockerfile (hunk 2 above)
Comment (arnout) [reply tools/docker/builder/Dockerfile:35]: Since the script will tell you to install uuidgen if you don't have it, I don't think it's needed to add it to the main readme. Anyway, nobody outside of the maintainer team is going to use that script.

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -83,6 +88,30 @@ build-in-docker-static:
   variables:
     EXTRA_CMAKE_FLAGS: "-DBUILD_SHARED_LIBS=OFF"
 
+upload-artifacts:
+  stage: upload
+  extends: .in-prplmesh-builder
+  script:
+    # the owncloud password is stored in the file pointed by "$CI_OWNCLOUD_PASSWORD".
+    # We can only store the password in it, since spaces are not supported in Gitlab masked variables.
+    # We symlink the file to .netrc and pre-pend to it to avoid manipulating the password itself
```
Comment (arnout) [inline .gitlab-ci.yml:97]: I still don't understand why we symlink instead of copy the file...

File: .gitlab-ci.yml [hunk 6]
```diff
@@ -83,6 +88,30 @@ build-in-docker-static:
...
+    - ln -s "$CI_OWNCLOUD_PASSWORD" ~/.netrc
+    - sed --follow-symlinks -i 's/^/machine ftp.essensium.com login prplmesh-robot-ci password /' ~/.netrc
+    - ci/owncloud/upload_to_owncloud.sh -v "artifacts/$CI_COMMIT_SHA-$CI_JOB_ID" ./build
+    - |
+      if [ $CI_COMMIT_BRANCH = "master" ] ; then
+          echo "Updating the 'latest' folder"
+          # Copying takes a lot of time, and we want the "latest" folder to be updated
+          # "as fast as possible", so we make a copy and move it:
+          ci/owncloud/owncloud_definitions.sh copy "artifacts/$CI_COMMIT_SHA-$CI_JOB_ID" "temp_uploads/latest_$CI_COMMIT_SHA-$CI_JOB_ID"
+          ci/owncloud/owncloud_definitions.sh move "temp_uploads/latest_$CI_COMMIT_SHA-$CI_JOB_ID" "artifacts/latest"
```
Comment (arnout) [inline .gitlab-ci.yml:108]: Nice!

**Review by arnout — APPROVED** [review]
Let's get this merged quickly.

File: ci/upload_to_owncloud.sh (hunk 3 above)
Comment (rmelotte) [reply ci/upload_to_owncloud.sh:61]: With a link, we would be able to make it point to something that is valid at any point in time, so it won't matter much anymore if destination folder already exist or not.

I agree it should be good enough for now.

File: .gitlab-ci.yml (hunk 5 above)
Comment (rmelotte) [reply .gitlab-ci.yml:97]: There is no particular reason for a symlink over a copy. 
If it feels more "natural", I can make a copy instead.
Is there a particular advantage to making a copy?

File: ci/upload_to_owncloud.sh (hunk 1 above)
Comment (rmelotte) [reply ci/upload_to_owncloud.sh]: @tomereli Like I said in another comment if the target directory does not exist, using a temporary directory shouldn't make a noticeable difference.
Did you test it on your side?
Do you still want to revisit it?
If not I'll add the "ready for merge" tag.

Comment (tomereli) [reply ci/upload_to_owncloud.sh]: Go ahead please, if we will need to, we will revisit this in the future.
